### PR TITLE
otp: Explain the term equivalence operators

### PR DESCRIPTION
**Pull request by jhogberg**
This PR attempts to document the term equivalence operators, whose behavior is [set to change slightly in OTP 27](https://erlangforums.com/t/in-erlang-otp-27-0-0-will-no-longer-be-exactly-equal-to-0-0/2586). It's a pretty subtle (and quite pedantic) change so I'd really appreciate all the feedback I can get, especially from the community at large.

### CODE REVIEW

File: system/doc/reference_manual/expressions.xml [hunk 1]
```diff
@@ -845,6 +859,8 @@ number &lt; atom &lt; reference &lt; fun &lt; port &lt; pid &lt; tuple &lt; map
 true
 2> <input>1=:=1.0.</input>
 false
+2> <input>+0.0=:=-0.0.</input>
+false
```
Comment (RaimoNiskanen) [inline system/doc/reference_manual/expressions.xml:863]: Should we have `0.0 =:= -0.0` and `+0.0 =:= 0.0`?

File: system/doc/reference_manual/expressions.xml [hunk 2]
```diff
@@ -836,6 +836,20 @@ number &lt; atom &lt; reference &lt; fun &lt; port &lt; pid &lt; tuple &lt; map
...
+      floats or integers (<c>1 == 1.0</c> is true), the term equivalence
+      operators return whether there exists any function that can tell their
+      arguments apart.</p>
+    <p>For example, while the terms <c>0.0</c> and <c>-0.0</c> represent the
+      exact same <em>number</em>, they yield different results when converted
+      to string form through <c>float_to_list/1</c>, so <c>=:=</c> and
+      <c>=/=</c> consider them different. This is useful when reasoning about
+      terms as opaque values, for example in associative containers or memoized
+      functions where using the equal-to (<c>==</c>) operator can result in
+      subtly incorrect results.</p>
```
Comment (RaimoNiskanen) [inline system/doc/reference_manual/expressions.xml:851]: To me this explanation explains why `==` and `=:=` behaves differently on `0.0` and `0`, but not why `=:=` regards `0.0` as different from `-0.0`.

Comment (RaimoNiskanen) [reply system/doc/reference_manual/expressions.xml:851]: Where should we explain that `0.0` and `+0.0` are exactly equal in runtime but have slightly different meaning in compile time?

Comment (jhogberg) [reply system/doc/reference_manual/expressions.xml:852]: >  To me this explanation explains why == and =:= behaves differently on 0.0 and 0, but not why =:= regards 0.0 as different from -0.0.

I've pushed up a commit that I hope will be a bit clearer.

> Where should we explain that 0.0 and +0.0 are exactly equal in runtime but have slightly different meaning in compile time?

I'm not sure that's necessary: they are the same even in compile-time, the compiler just happens to give the user a helpful warning when using the former in a pattern or term equality test.

File: system/doc/reference_manual/expressions.xml [hunk 3]
```diff
@@ -836,6 +836,21 @@ number &lt; atom &lt; reference &lt; fun &lt; port &lt; pid &lt; tuple &lt; map
       depending on the size of the float because otherwise comparison of large
       floats and integers would lose their transitivity.</p>
 
+    <p>The term equivalence operators, <c>=:=</c> and <c>=/=</c>, return
+      whether two terms are indistinguishable. While the other operators
+      consider the same <em>numbers</em> equal regardless of whether they are
```
Comment (bjorng) [inline system/doc/reference_manual/expressions.xml:841]: I think that "regardless of" can be omitted. At least, that sounds good to me.

**Review by bjorng — APPROVED** [review]

File: system/doc/reference_manual/expressions.xml (hunk 3 above)
Comment (jhogberg) [reply system/doc/reference_manual/expressions.xml]: Thanks, I've tweaked the wording a bit. :)

File: system/doc/reference_manual/expressions.xml [hunk 4]
```diff
@@ -836,6 +836,21 @@ number &lt; atom &lt; reference &lt; fun &lt; port &lt; pid &lt; tuple &lt; map
       depending on the size of the float because otherwise comparison of large
       floats and integers would lose their transitivity.</p>
 
+    <p>The term equivalence operators, <c>=:=</c> and <c>=/=</c>, return
+      whether two terms are indistinguishable. While the other operators
+      consider the same <em>numbers</em> equal even when their types differ
+      (<c>1 == 1.0</c> is true), the term equivalence operators return whether
+      there exists any function that can tell their arguments apart.</p>
+    <p>For example, while the terms <c>0.0</c> and <c>-0.0</c> represent the
+      exact same <em>number</em>, they yield different results when converted
```
Comment (RaimoNiskanen) [inline system/doc/reference_manual/expressions.xml:845]: I think we need a *since* note of some kind here since that distinction is new.

Comment (jhogberg) [reply system/doc/reference_manual/expressions.xml:845]: Good point! :)

Comment (RaimoNiskanen) [reply system/doc/reference_manual/expressions.xml:845]: We should also explain that `0.0` and `-0.0` are distinct terms here:
https://www.erlang.org/doc/reference_manual/data_types#representation-of-floating-point-numbers
with a note when that changed...

File: system/doc/reference_manual/expressions.xml (hunk 2 above)
Comment (RaimoNiskanen) [reply system/doc/reference_manual/expressions.xml]: But we need to state that they are the same

File: system/doc/reference_manual/expressions.xml (hunk 4 above)
Comment (RaimoNiskanen) [reply system/doc/reference_manual/expressions.xml:845]: Perhaps a good place to state that `0.0` and `+0.0` are equivalent

File: system/doc/reference_manual/expressions.xml (hunk 2 above)
Comment (jhogberg) [reply system/doc/reference_manual/expressions.xml]: Sure, I've pushed up another commit to that effect. :)

File: system/doc/reference_manual/expressions.xml [hunk 5]
```diff
@@ -836,6 +836,21 @@ number &lt; atom &lt; reference &lt; fun &lt; port &lt; pid &lt; tuple &lt; map
...
+      to string form through <c>float_to_list/1</c>: when given the former it
+      returns a string without a sign, and when given the latter it returns a
+      string with a sign. Therefore <c>=:=</c> and <c>=/=</c> consider these
+      terms different. This is useful when reasoning about terms as opaque
+      values, for example in associative containers or memoized functions where
```
Comment (RaimoNiskanen) [inline system/doc/reference_manual/expressions.xml:850]: I still think it becomes confusing here, maybe partly because I am familiar with the history.

Above the text talks about `0.0` and `-0.0` and below it talks about `1` vs. `1.0`.
They should not be in the same paragraph.

I suggest to paragraph break here, and add to the paragraph above that also `0 =/= 0.0` so `0 =/= 0.0 andalso 0.0 =/= -0.0 andaso -0.0 =/= 0` is `true`, that is: all these 3 zeros are different.
Then the paragraph below can start with "The equivalence operators are useful when reasoning ...".

I think we should not mix up the new distinction between `0.0` and `-0.0` with the old distinction between `0` and `0.0`.

The Note further down about the "bug" in the equivalence between `0.0` and `-=0.0` makes things clear.  I like that.

File: system/doc/reference_manual/data_types.xml [hunk 6]
```diff
@@ -87,6 +87,19 @@
...
+      <title>Comparisons</title>
+      <p>Both integers and floats share the same linear order. That is,
+        <c>1</c> compares less than <c>2.4</c>, <c>3</c> compares greater than
+        <c>2.99999</c>, and <c>5</c> is equal to <c>5.0</c>.</p>
+      <p>When wanting to match a specific number without also matching its
+        integer- or floating-point twin, the term equivalence operators
+        (<c>=:=</c>, <c>=/=</c>) should be used instead.</p>
+      <p>Note that since OTP 27, <c>0.0</c> and <c>-0.0</c> are <em>not</em>
+        considered to be the same term since they can be distinguished through
+        the use of string or bitstring conversion functions.</p>
```
Comment (RaimoNiskanen) [inline system/doc/reference_manual/data_types.xml:101]: At the end of the next section it is stated that we don not follow IEEE completely, and that Nan, +Inf and -Inf are not implemented in Erlang.  I suggest writing there that 0.0 vs. -0.0 *is* since it is also an IEEE feature, but one that we *do* support.

File: system/doc/reference_manual/data_types.xml [hunk 7]
```diff
@@ -87,6 +87,37 @@
...
 1234.333333
 </pre>
+<marker id="numeric_comparisons" />
+    <section>
+      <title>Comparisons</title>
+      <p>Both integers and floats share the same linear order. That is,
+        <c>1</c> compares less than <c>2.4</c>, <c>3</c> compares greater than
+        <c>2.99999</c>, and <c>5</c> is equal to <c>5.0</c>.</p>
+      <p>When wanting to match a specific number without also matching its
+        integer- or floating-point twin, it may be tempting to use the term
```
Comment (josevalim) [inline system/doc/reference_manual/data_types.xml:97]: Nitpick: the negation is hard to process here, a suggestion to special out the specific comparison pairs:

```suggestion
      <p>When wanting to match an integer with another integer or a float
        with another float, it may be tempting to use the term
```

File: system/doc/reference_manual/data_types.xml [hunk 8]
```diff
@@ -87,6 +87,37 @@
...
+        equivalence operators (<c>=:=</c>, <c>=/=</c>) or pattern matching.
+        This works for integers which has a distinct representation for every
+        number, but does not work for floating-point as the latter has two
+        representations for zero which are considered different by the term
+        equivalence operator (<c>=:=</c>).</p>
+      <p>If you wish to specifically compare floating-point numbers, use the
+        regular comparison operators (such as <c>==</c>) and add guards that
+        require both the arguments to be floating-point.</p>
```
Comment (josevalim) [inline system/doc/reference_manual/data_types.xml:105]: Nitpick: I am not sure if we should label it as "does not work". As you said later, you are comparing them as terms, not as numbers, which is fine by IEEE754. To me this is 

```suggestion
        However, beware floating-point numbers have two representations for
        zero (<c>0.0</c> and <c>-0.0</c>) and the term equivalent operator
        (<c>=:=</c>) will consider them to be distinct, which may be undesired.
        This is not a concern for integers, which have a distinct representation
        for every number.</p>
      <p>If you wish to specifically compare floating-point numbers while
        keeping the zero equivalence, use the regular comparison operators
        (such as <c>==</c>) and add guards that require both the arguments
        to be floating-point.</p>
```

Comment (jhogberg) [reply system/doc/reference_manual/data_types.xml]: Thanks, I've reworded it a bit :)

File: system/doc/reference_manual/data_types.xml [hunk 9]
```diff
@@ -87,6 +87,37 @@
 11> <input>1_234.333_333</input>
 1234.333333
 </pre>
+<marker id="numeric_comparisons" />
+    <section>
+      <title>Comparisons</title>
+      <p>Both integers and floats share the same linear order. That is,
+        <c>1</c> compares less than <c>2.4</c>, <c>3</c> compares greater than
+        <c>2.99999</c>, and <c>5</c> is equal to <c>5.0</c>.</p>
+      <p>When wanting to an integer with another integer or a float with
```
Comment (Maria-12648430) [inline system/doc/reference_manual/data_types.xml:96]:
```suggestion
      <p>When wanting to compare an integer with another integer or a float with
```

Comment (jhogberg) [reply system/doc/reference_manual/data_types.xml]: Thanks :)

File: system/doc/reference_manual/data_types.xml [hunk 10]
```diff
@@ -87,6 +87,37 @@
 11> <input>1_234.333_333</input>
 1234.333333
 </pre>
+<marker id="numeric_comparisons" />
+    <section>
+      <title>Comparisons</title>
+      <p>Both integers and floats share the same linear order. That is,
+        <c>1</c> compares less than <c>2.4</c>, <c>3</c> compares greater than
+        <c>2.99999</c>, and <c>5</c> is equal to <c>5.0</c>.</p>
+      <p>When wanting compare an integer with another integer or a float with
```
Comment (Maria-12648430) [inline system/doc/reference_manual/data_types.xml:96]:
```suggestion
      <p>When wanting to compare an integer with another integer or a float with
```
:sweat_smile::sweat_smile::sweat_smile:

Comment (Maria-12648430) [reply system/doc/reference_manual/data_types.xml:96]: On another note, in my previous comment I suggested "match", but you used "compare", which is certainly the better choice here.
But since you're subsequently talking only about testing for (in-)equality, it may be good to end the sentence with something like "... to check for equality or inequality", or write it like "... to use pattern matching, or the term equivalence operators <c>=:=</c> to test for equality and <c>=/=</c> to test for inequality"?

Comment (jhogberg) [reply system/doc/reference_manual/data_types.xml:96]: I fear that'll just make things too wordy, it's pretty bad as it is. :thinking:

Comment (Maria-12648430) [reply system/doc/reference_manual/data_types.xml]: Yeah, it is pretty hard to be concise _and_ precise :sweat: I guess the message gets across even without the equality/inequality part, so it's fine as far as I'm concerned.

File: system/doc/reference_manual/expressions.xml [hunk 11]
```diff
@@ -836,6 +836,28 @@ number &lt; atom &lt; reference &lt; fun &lt; port &lt; pid &lt; tuple &lt; map
...
+      same <em>number</em>, but they yield different results when converted to
+      string form through <c>float_to_list/1</c>: when given the former it
+      returns a string without a sign, and when given the latter it returns a
+      string with a sign. Therefore, <c>=:=</c> and <c>=/=</c> consider
+      them different.</p>
+
+    <p>The term equivalence operators are useful when reasoning about terms as 
+      opaque values, for example in associative containers or memoized
+      functions where using the equal-to operator (<c>==</c>) can result in
+      subtly incorrect results.</p>
```
Comment (Maria-12648430) [inline system/doc/reference_manual/expressions.xml:859]: This is also somewhat confusing. It brings up two new concepts, without explanation or examples (which of course would blow this document out of all proportion). I mean, I know about the subtly incorrect results you mention here, because they did bite me in the past. But I'd say that I wouldn't have been aware of them if I had read this paragraph before :woman_shrugging:

Maybe you should just show briefly how different implementations of the same `$THING` behave differently by using the equality vs term equivalence operators/pattern matching, like `ordsets` (working by equality) and `sets` (working by term equivalence), or the like?
As of late, there is a NOTE box at the top of the [`sets` documentation](https://www.erlang.org/doc/man/sets.html) outlining the differences in behavior between `sets`, `ordsets` and `gb_sets` that you could link to.

Also, I think it is wrong to unconditionally say that using the equality operators can result in incorrect results, it kinda leaves an aftertaste of "equality operators ***bad***!!!". Which they are not: If equality is what you _want_, I'd say it is not wrong.
IMO the real issue is that there is the possibility of a mismatch between how you _expect_ something to behave and how it _really_ behaves if you don't pay attention to equality vs term equivalence, a mismatch that will probably manifest itself every so often, and is hard to find.

Comment (Maria-12648430) [reply system/doc/reference_manual/expressions.xml:859]: Reading back on my comment, I realize it sounds a bit harsh :sweat: Sorry for that :pray:

But I realize the wording and maybe-needed verbosity depends to some degree on what assumptions we can make about prior knowledge of a reader.

**Review by mikpe — APPROVED** [review]

File: system/doc/reference_manual/expressions.xml [hunk 12]
```diff
@@ -836,6 +836,28 @@ number &lt; atom &lt; reference &lt; fun &lt; port &lt; pid &lt; tuple &lt; map
       depending on the size of the float because otherwise comparison of large
       floats and integers would lose their transitivity.</p>
 
+    <p>The term equivalence operators, <c>=:=</c> and <c>=/=</c>, return
+      whether two terms are indistinguishable. While the other operators
+      consider the same <em>numbers</em> equal even when their types differ
+      (<c>1 == 1.0</c> is true), the term equivalence operators return whether
+      there exists any function that can tell their arguments apart.</p>
```
[the remark this answers is not in the document]
Comment (jhogberg) [reply system/doc/reference_manual/expressions.xml:843]: `∀A∀B ¬(A =/= B) ≡ (A =:= B) ≡ ∀f (f(A) =:= f(B)))`, what's the problem? :grin:

> "... whether there exists **no** function...", 

The operators are mirrors of each other, so they both say the same thing, and I'm not sure the extra negation helps. :thinking: 

> Maybe just stick with the conventional notion of "... are exactly the same...", 

Hacker News got all up in arms because they misunderstood what we meant by "exactly equal to," so I'm not sure that'll help much either. :sweat_smile:

File: system/doc/reference_manual/expressions.xml (hunk 11 above)
Comment (jhogberg) [reply system/doc/reference_manual/expressions.xml:859]: > But I realize the wording and maybe-needed verbosity depends to some degree on what assumptions we can make about prior knowledge of a reader.

If only we could assume that the reader knew everything to begin with... :smile: 

> it kinda leaves an aftertaste of "equality operators bad!!!". Which they are not: If equality is what you want, I'd say it is not wrong.

It's wrong when reasoning about terms as opaque values, which is what that paragraph is about.

Can you try playing around with the wording to see if there's any particular direction you'd prefer (in general, both files if you wish)?

File: system/doc/reference_manual/expressions.xml (hunk 12 above)
[the remark this answers is not in the document]
Comment (josevalim) [reply system/doc/reference_manual/expressions.xml:843]: I definitely prefer term-equivalence operators as a name. I took a stab at writing this section. Feel free use any part of it or throw it all out. The idea is that we would describe the general rules, as they are currently written, provide examples, and then discuss comparing numbers in their own section.

---

The rules above apply to all comparison operators. The distinction from `==` and `/=` to `=:=` and `=/=` arises when comparing integers and floats. Let's explore this case in more detail.

## Comparing numbers

Similar to `+` and other arithmetic operators, Erlang provides a series of equality operators to compare numbers, regardless if they are integers or floats:

```erlang
1> 1 == 1.0.
true
2> 1 /= 1.0.
false
3> 10 < 5.0.
false
4> 10 >= 5.
true
```

When comparing an integer to a float, the number with the lesser precision is converted into the type of the other. A float is more precise than an integer until all significant figures of the float are to the left of the decimal point. This happens when the float is larger/smaller than `+/-9007199254740992.0`. The conversion strategy is changed depending on the size of the float because otherwise comparison of large floats and integers would lose their transitivity.

Additionally, floating-point numbers have an interesting property where zero is either negative (-0.0) or positive (+0.0), although in practice they compare equal:

```erlang
1> +0.0 == -0.0
true
```

However, in some cases, you may want to distinguish between integers and floats, or between positive and negative zero, as they have different in-memory representations (and in some mathematical domains they may also exhibit different properties). Erlang provides two _term equivalence operators_ (`=:=` and `=/=`) for this exact purpose:

```erlang
1> 1 =:= 1.0.
false
2> +0.0 =:= -0.0
false
```

Finally, note pattern matching relies on term equivalence. Therefore, the following expressions won't match either:

```erlang
1> 1 = 1.0.
** exception error: no match of right hand side value 1.0
2> -0.0 = +0.0.
```

To avoid ambiguity, Erlang will warn if 0.0 is used inside a pattern. You must instead match on `+0.0` or `-0.0`. If you want to match on both positive and negative zero floating points, you can write:

```erlang
is_floating_zero(Num) when is_float(Num), Num == 0.0 -> true;
is_floating_zero(_Num) -> false.
```

In other words, the comparison operators will behave in accordance to the other arithemetic operators in the language, and adhere to the IEEE 754 specification. The term equivalence operators allow developers to explicitly compare their underlying representations.

> Note: Prior to OTP 27, the term equivalence operators had a bug where they considered <c>0.0</c> and <c>-0.0</c> to be the same term.

[the remark this answers is not in the document]
Comment (josevalim) [reply system/doc/reference_manual/expressions.xml:843]: FWIW, Elixir calls those operators "strictly equal". It may make more sense to think some comparisons are more strict than others (rather than more exact).

[the remark this answers is not in the document]
Comment (jhogberg) [reply system/doc/reference_manual/expressions.xml:843]: Thanks, I like that take on things. I'll try to incorporate it when I'm back from vacation mid-to-late September.

File: system/doc/reference_manual/expressions.xml (hunk 11 above)
Comment (Maria-12648430) [reply system/doc/reference_manual/expressions.xml:859]: > Can you try playing around with the wording to see if there's any particular direction you'd prefer (in general, both files if you wish)?

I guess I had that coming :rofl: Ok, I'll think and play :wink:

**Comment by jhogberg** [comment]
I've changed the wording a slight bit, keeping most of the PR the same. I'd like to continue in the direction @josevalim took but I think that's better done in a general overhaul of this subject as it's currently split between "expressions" and "data types." I think a separate PR will be better for that.

If that sounds good, I'll merge this PR next week and start work on implementing the comparison change itself.
